File: binary-search/find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.go
package findfirstandlastpositionofelementinsortedarray

func searchRange(nums []int, target int) []int {
	left := 0
	right := len(nums) - 1

	ans := []int{-1, -1}

	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] > target || nums[mid] == target {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}

	if left == len(nums) || nums[left] != target {
		return ans
	}

	ans[0] = left

	right = len(nums) - 1

	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] > target {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}

	ans[1] = right

	return ans
}

File: binary-search/find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
function searchRange(nums, target) {
  let left = 0;
  let right = nums.length - 1;

  const ans = [-1, -1];

  while (left <= right) {
    let mid = left + Math.floor((right - left) / 2);

    if (nums[mid] > target || nums[mid] === target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  if (left === nums.length || nums[left] !== target) {
    return ans;
  }

  ans[0] = left;

  right = nums.length - 1;

  while (left <= right) {
    let mid = left + Math.floor((right - left) / 2);

    if (nums[mid] > target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  ans[1] = right;

  return ans;
}
